contributions file (#235)
equinor/esv-intersection: CONTRIBUTION.md

Contributing changes
====================

Licensing of contributed material
---------------------------------

Keep in mind as you contribute, that code, docs and other material submitted to
open source projects are usually considered licensed under the same terms
as the rest of the work.

The details vary from project to project, but from the perspective of this
document's authors:

- Anything submitted to a project falls under the licensing terms in the
  repository's top level ``LICENSE`` file.

    - For example, if a project's ``LICENSE`` is BSD-based, contributors should
      be comfortable with their work potentially being distributed in binary
      form without the original source code.

- Per-file copyright/license headers are typically extraneous and undesirable.
  Please don't add your own copyright headers to new files unless the project's
  license actually requires them!

    - Not least because even a new file created by one individual (who often
      feels compelled to put their personal copyright notice at the top) will
      inherently end up contributed to by dozens of others over time, making a
      per-file header outdated/misleading.

> Version control branching
> -------------------------
> 
> * Always **fork the repo** or **make a new branch** for your work, no matter how small. This makes
>   it easy for others to take just that one set of changes from your repository,
>   in case you have multiple unrelated changes floating around.
> 
>     * A corollary: **don't submit unrelated changes in the same branch/pull
>       request**! The maintainer shouldn't have to reject your awesome bugfix
>       because the feature you put in with it needs more review.
> 
> * **Base your new branch off of the appropriate branch** on the main
>   repository:
> 
>     * **Bug fixes** should be based on the branch named after the **oldest
>       supported release line** the bug affects.
> 
>         * E.g. if a feature was introduced in 1.1, the latest release line is
>           1.3, and a bug is found in that feature - make your branch based on
>           1.1.  The maintainer will then forward-port it to 1.3 and master.
>         * Bug fixes requiring large changes to the code or which have a chance
>           of being otherwise disruptive, may need to base off of **master**
>           instead. This is a judgement call -- ask the devs!
> 
>     * **New features** should branch off of **the 'master' branch**.
> 
>         * Note that depending on how long it takes for the dev team to merge
>           your patch, the copy of ``master`` you worked off of may get out of
>           date! If you find yourself 'bumping' a pull request that's been
>           sidelined for a while, **make sure you rebase or merge to latest
>           master** to ensure a speedier resolution.

Reviewer: ```suggestion
Version control branching
-------------------------

There are 2 active branches on this repo, plus branches for each major version (see below):
- The `master` branch is the main branch of the repository, the one that is used for active development
  - Every time this branch is modified (typically when merging a PR), a Storybook is automatically deployed (on _Radix_, `master` environment). It can be found [here](TODO: add link)
  - Every time this branch is modified, an updated package is deployed to [npm](TODO: add link)
- The `latest` branch always points at the the most recent version released
  - The branch has to be updated manually, whenever a new release is created
  - Every time this branch is updated, the Storybook is automatically deployed (on _Radix_, `latest` environment). It can be found [here](TODO: add link)

This repository relies on [Semantic Versioning](https://semver.org/).
 
In order to create a release for a new major version, the first step is to create a new branch named `version_X` in the repo (where `X` has to be replaced with an incremental number). Then, a new _Release_ is created from that branch here on Github, named `vX.0.0`. Remember to update the `latest` branch to the new release.

Development continues on the `master` branch. In order to create a new minor version, the desired changes have to be moved to the `version_X` branch (either by merging, rebasing, cherry-picking, …). After proper testing, a new _Release_ can be created here on Github, named `vX.Y.0`. Remember to update the `latest` branch to the new release.

If a hotfix has to be applied to a released version, the fix has to be added on top of `version_X`. Then a new _Release_ can be created here on Github, named `vX.Y.Z`. Remember to update the `latest` branch to the new release. If the fix is relevant also for `master`, make sure the fix is ported on top of `master` as well.

When you want to contribute to the repo, always **fork the repo** or **make a new branch** for your work, no matter how small. This makes it easy for others to take just that one set of changes from your repository, in case you have multiple unrelated changes floating around.

    * A corollary: **don't submit unrelated changes in the same branch/pull request**!

* **Base your new branch off of the appropriate branch** on the main repository:

    * **Bug fixes** should be based either on `master` or on `latest` branches. Contribution to an older `version_X` branch (i.e., any version branch that does not correspond with `latest`) will be **rejected** (there can be exceptions, but that would require very compelling arguments). 

        * Bug fixes requiring large changes to the code or which have a chance
          of being otherwise disruptive, may need to based on top of **master**.
          This is a judgement call -- ask the devs!

    * **New features** should branch off of the `master` branch

        * Note that depending on how long it takes for the dev team to merge
          your patch, the copy of ``master`` you worked off of may get out of
          date! If you find yourself 'bumping' a pull request that's been
          sidelined for a while, **make sure you rebase or merge to latest
          master** to ensure a speedier resolution.
```

Author: i think this commit suggestion would be better as a pull request into this branch, I would like to see some of it changed: "  - Every time this branch is modified, an updated package is deployed to [npm](TODO: add link)", we don't have a fully automatic deployment to npm, i'm also not sure if we should.

We should also describe our deployment process, that we need to bump the version in  `package-lock` and `package.json`. That we create a release in GitHub that triggers an action that pushes to npm. And that the newest stuff in the changelog would be a part of the details for the release

Reviewer: Sounds reasonable. Let's merge this then, and we will consider further changes in dedicated PRs

4. Hit 'submit'! And please be patient - the maintainers will get to you when
   they can.

## Setup guide

### Cloning repository

Start by cloning the repository to desired directory.

```
git clone https://github.com/equinor/esv-intersection.git
```

### Install dependencies

Install all dev dependencies defined in package.json using node.

```
npm install
```

## Usage

### Storybook

```
npm run storybook
```

### Document generation

```
npm run docs
```

This command will overwrite any old documentation

### Creating tests

All tests are defined within the test-folder. Jest naming convention is to place tests for SOMESCRIPT.ts in a single file SOMESCRIPT.test.ts.

### Testing

```
npm run test
```

Executes all tests defined within the test folder.

```
npm run test:watch
```

Executes all tests, but does not return immediately. Makes it possible to re-run failed tests quickly.

### Building/Compiling

```
npm run build
```

Compiles the code found within the src-folder. Build is outputted to a new dist-folder.

----
The content of this file is largely inspired by [bitprophet's contribution guide](https://github.com/bitprophet/contribution-guide.org/blob/master/index.rst).

> This file is licensed under the _BSD 2-Clause "Simplified" License_ that follows:

Reviewer: What if we move these two lines at the beginning of the file, and leave the license here at the end?

> 

Reviewer: ```suggestion
The content of this file is largely inspired by [bitprophet's contribution guide](https://github.com/bitprophet/contribution-guide.org/blob/master/index.rst).
This file is licensed under the _BSD 2-Clause "Simplified" License_ that follows:   

-------------
Copyright (c) 2020 Jeff Forcier.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-------------

```

# How to contribute

## Submitting bugs

### Due diligence
-------------

> Before submitting a bug, please do the following:

Author: Sure, this would be redundant anyway considering we have a bug template when creating issues. We should make sure that the template there is easily understood and easy to use

  license actually requires them!

    - Not least because even a new file created by one individual (who often
      feels compelled to put their personal copyright notice at the top) will
      inherently end up contributed to by dozens of others over time, making a
      per-file header outdated/misleading.

Version control branching
-------------------------

* Always **make a new branch** for your work, no matter how small. This makes
  it easy for others to take just that one set of changes from your repository,
  in case you have multiple unrelated changes floating around.

    * A corollary: **don't submit unrelated changes in the same branch/pull
      request**! The maintainer shouldn't have to reject your awesome bugfix
      because the feature you put in with it needs more review.

* **Base your new branch off of the appropriate branch** on the main
  repository:

    * **Bug fixes** should be based on the branch named after the **oldest
      supported release line** the bug affects.

        * E.g. if a feature was introduced in 1.1, the latest release line is
          1.3, and a bug is found in that feature - make your branch based on

>           1.1.  The maintainer will then forward-port it to 1.3 and master.

Author: this section needs to be reworked according to how we want to operate. At the moment we've only used the master branch, and create releases off of what has changed since then. We should decide on a clear plan of how we want to control features and bugfixes going into a release.

/**
 * increment number by 1
 * @param n - number to be incremented
 */
increment(n) {
  return n + 1;
}
```

Tests aren't optional
---------------------

Any bugfix that doesn't include a test proving the existence of the bug being
fixed, may be suspect.  Ditto for new features that can't prove they actually
work.

Your ideally include unit tests and/or storybooks to test and help describe what you are solving

We've found that test-first development really helps make features better
architected and identifies potential edge cases earlier instead of later.
Writing tests before the implementation is strongly encouraged.

Full example
------------

Here's an example workflow for a project ``theproject`` hosted on Github, which
is currently in version 1.3.x. Your username is ``yourname`` and you're
submitting a basic bugfix.

Preparing your Fork
^^^^^^^^^^^^^^^^^^^

1. Click 'Fork' on Github, creating e.g. ``yourname/theproject``.
2. Clone your project: ``git clone git@github.com:yourname/theproject``.
3. ``cd theproject``
4. Install the development requirements: ``npm install``.
5. Create a branch: ``git checkout -b foo-the-bars 1.3``.

Making your Changes
^^^^^^^^^^^^^^^^^^^

> 1. Add changelog entry crediting yourself.

Author: No, I currently use release notes and look over pull requests. It is a decent way to keep track of changes, but I think it would be easier to look at the changelog instead of pr history

   an existing issue) or describe the issue + your fix (if submitting a wholly
   new bugfix).
4. Hit 'submit'! And please be patient - the maintainers will get to you when
   they can.

## Setup guide

### Cloning repository

Start by cloning the repository to desired directory.

```
git clone https://github.com/equinor/esv-intersection.git
```

### Install dependencies

Install all dev dependencies defined in package.json using node.

```
npm install
```

## Usage

### Storybook

```
npm run storybook
```

### Document generation

```
npm run docs
```

This command will overwrite any old documentation

### Creating tests

All tests are defined within the test-folder. Jest naming convention is to place tests for SOMESCRIPT.ts in a single file SOMESCRIPT.test.ts.

### Testing

```
npm run test
```

Executes all tests defined within the test folder.

```
npm run test:watch
```

Executes all tests, but does not return immediately. Makes it possible to re-run failed tests quickly.

### Building/Compiling

```
npm run build
```

Compiles the code found within the src-folder. Build is outputted to a new dist-folder.

> Thanks to [bitprophet](https://github.com/bitprophet/contribution-guide.org/blob/master/index.rst) for contribution guide

Reviewer: ```suggestion
```

# How to contribute

Pull requests should include the following:

> - link to an issue that describing the problem, or a description in the pull request describing the issue and what it solves

Reviewer: Typo, either `that is describing` or `that describes`

> - link to an issue that describing the problem, or a description in the pull request describing the issue and what it solves

Author: ```suggestion
- link to an issue that describes the problem, or a description in the pull request describing the issue and what it solves
```

- unit tests (if applicable)
- storybook (if applicable)
- a separate commit that contains the ouput from running `npm run docs`

## Setup guide

### Cloning repository

Start by cloning the repository to desired directory.

```
git clone https://github.com/equinor/esv-intersection.git
```

### Install dependencies

Install all dev dependencies defined in package.json using node.

```
npm install
```

## Usage

### Storybook

```
npm run storybook
```

### Document generation

```
npm run docs
```

> Three part process:

Reviewer: This is a bit unclear. What is the exact relation between `npm run docs` above, and the three part process described here?

Author: the three part process describes what happens when running the script command. it's an over-explain and we can shorten it to something like: 'this command will overwrite the old documentation'

> 1. Deletes the docs-folder, if it exists.

Reviewer: If the folder is actually called docs, I'd write 
```
Delete the `docs` folder
```

Author: you can use ```suggestion at the start of the formatting, while selecting the line you want to change. this way the suggestion can be commited directly from here

Author: ```suggestion
1. Deletes the `docs` folder, if it exists.
```

> Three part process:
> 
> 1. Deletes the docs-folder, if it exists.
> 2. Automatically generates documentation to a new docs-folder.
> 3. Copies the images-folder and .nojekyll into docs.
> 
> The copying of images makes it possible to refer to local images within the README!
> 
> The empty .nojekyll file makes it possible to upload html pages starting wth underscore to GitHub pages.

Author: ```suggestion
This command will overwrite any old documentation
```
